src/ScheduleParam.h

```diff
@@ -0,0 +1,254 @@
+        scalar.set_scalar<T>(value);
+    }
+
+    template <typename T2, typename std::enable_if<std::is_same<T2, LoopLevel>::value>::type * = nullptr>
+    HALIDE_ALWAYS_INLINE void typed_setter_impl(const LoopLevel &value, const char *msg) {
+        user_assert(is_looplevel_param()) << "Only ScheduleParam<LoopLevel> can be set withLoopLevel.";
+        loop_level.copy_from(value);
+    }
+
+    template <typename T2, typename std::enable_if<!std::is_convertible<T2, T>::value>::type * = nullptr>
+    HALIDE_ALWAYS_INLINE void typed_setter_impl(const T2 &value, const char *type) {
+        user_error << "The ScheduleParam " << name() << " cannot be set with a value of type " << type << ".\n";
+    }
+
+    template <typename T2 = T,
+              typename std::enable_if<std::is_same<T2, LoopLevel>::value>::type * = nullptr>
+    NO_INLINE void set_from_string_impl(const std::string &new_value_string) {
+        if (new_value_string == "root") {
+            set(LoopLevel::root());
+        } else if (new_value_string == "inline") {
+            set(LoopLevel::inlined());
+        } else {
+            user_error << "Unable to parse " << name() << ": " << new_value_string;
+        }
+    }
+
+    template <typename T2 = T,
+              typename std::enable_if<std::is_same<T2, bool>::value>::type * = nullptr>
+    NO_INLINE void set_from_string_impl(const std::string &new_value_string) {
+        if (new_value_string == "true") {
+            set(true);
+        } else if (new_value_string == "false") {
+            set(false);
+        } else {
+            user_error << "Unable to parse " << name() << ": " << new_value_string;
+        }
+    }
+
+    template <typename T2 = T,
+              typename std::enable_if<std::is_arithmetic<T2>::value && !std::is_same<T2, bool>::value>::type * = nullptr>
+    NO_INLINE void set_from_string_impl(const std::string &new_value_string) {
+        std::istringstream iss(new_value_string);
+        T t;
+        iss >> t;
+        user_assert(!iss.fail() && iss.get() == EOF) << "Unable to parse " << name() << ": " << new_value_string;
+        set(t);
+    }
+
+    void set_from_string(const std::string &new_value_string) override {
+        set_from_string_impl(new_value_string);
+    }
+
+public:
+    using type = T;
+
+    ScheduleParam() : Internal::ScheduleParamBase(get_param_type()) {}
+
+    explicit ScheduleParam(const std::string &name) : Internal::ScheduleParamBase(get_param_type(), name) {}
+
+    ScheduleParam(const std::string &name, const T &value) : Internal::ScheduleParamBase(get_param_type(), name) {
+        set(value);
+    }
+
+    ScheduleParam(const std::string &name, const std::string &value) : Internal::ScheduleParamBase(get_param_type(), name) {
+        set_from_string(value);
+    }
+
+#define HALIDE_SCHEDULE_PARAM_TYPED_SETTER(TYPE) \
+    void set(const TYPE &new_value) override { typed_setter_impl<TYPE>(new_value, #TYPE); }
+
+    HALIDE_SCHEDULE_PARAM_TYPED_SETTER(bool)
```

Comment:
So: yes, I could eliminate all the extra set() stuff right now; it does exist to serve a hopefully-soon-future purpose: The goal they serve is to allow use of non-customized GeneratorStub in a way that is still type-safe at compile-time; this has been prototyped in https://github.com/halide/Halide/tree/srj-stub2, but there's no PR yet because the exact shape of it still needs work pending the outcome of ScheduleParam.

If it's perceived as a blocker for this I'm happy to simplify it away for now, but it's going to resurface again in the future (though maybe it will make more sense then)...